src/modules/kvs/kvs.c

```diff
@@ -2689,6 +2752,140 @@ static void setroot_unpause_request_cb (flux_t *h, flux_msg_handler_t *mh,
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int sync_config (struct kvs_ctx *ctx,
+                        const flux_conf_t *conf,
+                        char *errbuf,
+                        int errbufsize)
+{
+    double sync = ctx->sync;
+
+    if (sync_config_parse (ctx, conf, &sync, errbuf, errbufsize) < 0)
+        return -1;
+
+    if (ctx->rank == 0 && sync != ctx->sync) {
+
+        /* setup timer if it wasn't done before */
+        if (sync > 0.0 && !ctx->sync_w) {
+            flux_reactor_t *r = flux_get_reactor (ctx->h);
+            if (!(ctx->sync_w = flux_timer_watcher_create (r,
+                                                           sync,
+                                                           0.0,
+                                                           sync_cb,
+                                                           ctx))) {
+                snprintf (errbuf,
+                          errbufsize,
+                          "error creating sync timer: %s",
+                          flux_strerror (errno));
+                return -1;
+            }
+        }
+
+        /* Potential sync timer adjustments
+         *
+         * If sync txn already submitted, let it process and new sync
+         * timer will take place next time (or not take place if sync
+         * disabled).
+         *
+         * If a timer is currently running, we take a simple approach.
+         * We stop the timer and restart it with the new time (or
+         * disable it if the timer is disabled).
+         *
+         * If no timer running, we'll just wait for next time a kvs
+         * transaction completes and let start_sync_timer() handle it.
+         *
+         * However, we make one minor exception, if the sync was
+         * disabled before and atleast one commit has occurred, we'll
+         * start the sync timer.
+         */
+        if (!ctx->sync_txn_submitted) {
+            if (ctx->sync_timer_started) {
+                if (sync > 0.0) {
+                    flux_watcher_stop (ctx->sync_w);
+                    flux_timer_watcher_reset (ctx->sync_w, sync, 0.);
+                    flux_watcher_start (ctx->sync_w);
+                }
+                else { /* sync == 0.0 */
+                    flux_watcher_stop (ctx->sync_w);
+                    ctx->sync_timer_started = false;
+                }
+            }
+            else {
+                if (sync > 0.0
+                    && ctx->sync == 0.0
+                    && ctx->primary_commit_since_last_sync) {
+                    /* on rank 0, primary namespace has to be here */
+                    struct kvsroot *root;
+                    root = kvsroot_mgr_lookup_root (ctx->krm,
+                                                    KVS_PRIMARY_NAMESPACE);
+                    assert (root);
+                    run_sync_timer (ctx);
+                }
+            }
+        }
+
+        ctx->sync = sync;
```

Comment:
not sure I can simplify it that much, but you're right that it could be simplified.  The above did have some dead code (`kvsroot_mgr_lookup_root()` looks up something that is not used).  If I used `start_sync_timer_if_needed()`, I'd have to do the lookup.

I did clean it up to the following.

```
    if (ctx->rank == 0 && sync != ctx->sync) {
        ctx->sync = sync;
        if (!ctx->sync_txn_submitted) {
            if (ctx->sync_timer_started) {
                flux_watcher_stop (ctx->sync_w);
                if (ctx->sync > 0.0) {
                    flux_timer_watcher_reset (ctx->sync_w, ctx->sync, 0.);
                    flux_watcher_start (ctx->sync_w);
                }
                else /* ctx->sync == 0.0 */
                    ctx->sync_timer_started = false;
            }
            else {
                if (ctx->primary_commit_since_last_sync)
                    start_sync_timer (ctx);
            }
        }
    }
```